File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/BluetoothDeviceFragment.java
package com.kinwatt.powermeter.ui.fragments;

import android.bluetooth.BluetoothDevice;
import android.content.Context;
import android.os.Bundle;
import android.os.ParcelUuid;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.kinwatt.powermeter.R;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BluetoothDeviceFragment extends Fragment {

    private OnListFragmentInteractionListener listener;

    private BluetoothDeviceRecyclerViewAdapter adapter;

    private List<BluetoothDevice> items = new ArrayList<>();
    private Map<BluetoothDevice, List<ParcelUuid>> map = new HashMap<>();

    public BluetoothDeviceFragment() { }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_bluetoothdevice_list, container, false);

        RecyclerView recyclerView = (RecyclerView) view;
        recyclerView.setLayoutManager(new LinearLayoutManager(view.getContext()));
        recyclerView.setAdapter(adapter = new BluetoothDeviceRecyclerViewAdapter(items, map, listener));

        return view;
    }

    public boolean contains(BluetoothDevice device) {
        return items.contains(device);
    }

    public void add(BluetoothDevice device, List<ParcelUuid> uuids) {
        items.add(device);
        map.put(device, uuids);
        adapter.notifyItemInserted(items.size() - 1);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnListFragmentInteractionListener) {
            listener = (OnListFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString() + " must implement OnListFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }

    public interface OnListFragmentInteractionListener {
        void onListFragmentInteraction(BluetoothDevice device, List<ParcelUuid> uuids);
    }
}

File: app/src/main/java/com/kinwatt/powermeter/ui/fragments/BluetoothDeviceFragment.kt
package com.kinwatt.powermeter.ui.fragments

import android.bluetooth.BluetoothDevice
import android.content.Context
import android.os.Bundle
import android.os.ParcelUuid
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.kinwatt.powermeter.R
import com.kinwatt.powermeter.databinding.FragmentBluetoothdeviceBinding

import java.util.ArrayList
import java.util.HashMap

class BluetoothDeviceFragment : Fragment() {

    private lateinit var binding: FragmentBluetoothdeviceBinding
    private var listener: OnListFragmentInteractionListener? = null

    private val deviceAdapter: BluetoothDeviceRecyclerViewAdapter by lazy { BluetoothDeviceRecyclerViewAdapter(items, map, listener) }

    private val items = ArrayList<BluetoothDevice>()
    private val map = HashMap<BluetoothDevice, List<ParcelUuid>>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        binding = FragmentBluetoothdeviceBinding.inflate(inflater, container, false)
        val view = binding.root

        val recyclerView = view as RecyclerView
        recyclerView.layoutManager =
            LinearLayoutManager(recyclerView.context)
        recyclerView.adapter = deviceAdapter

        return view
    }

    operator fun contains(device: BluetoothDevice): Boolean {
        return items.contains(device)
    }

    fun add(device: BluetoothDevice, uuids: List<ParcelUuid>) {
        items.add(device)
        map[device] = uuids
        deviceAdapter.notifyItemInserted(items.size - 1)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)

        if (context is OnListFragmentInteractionListener) {
            listener = context
        } else {
            throw RuntimeException(context!!.toString() + " must implement OnListFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        listener = null
    }

    interface OnListFragmentInteractionListener {
        fun onListFragmentInteraction(device: BluetoothDevice, uuids: List<ParcelUuid>)
    }
}
